File: zh-Hant/codes/java/chapter_searching/binary_search_edge.java
/**
 * File: binary_search_edge.java
 * Created Time: 2023-08-04
 * Author: krahets (krahets@163.com)
 */

package chapter_searching;

public class binary_search_edge {
    /* 二分搜尋最左一個 target */
    static int binarySearchLeftEdge(int[] nums, int target) {
        // 等價於查詢 target 的插入點
        int i = binary_search_insertion.binarySearchInsertion(nums, target);
        // 未找到 target ，返回 -1
        if (i == nums.length || nums[i] != target) {
            return -1;
        }
        // 找到 target ，返回索引 i
        return i;
    }

    /* 二分搜尋最右一個 target */
    static int binarySearchRightEdge(int[] nums, int target) {
        // 轉化為查詢最左一個 target + 1
        int i = binary_search_insertion.binarySearchInsertion(nums, target + 1);
        // j 指向最右一個 target ，i 指向首個大於 target 的元素
        int j = i - 1;
        // 未找到 target ，返回 -1
        if (j == -1 || nums[j] != target) {
            return -1;
        }
        // 找到 target ，返回索引 j
        return j;
    }

    public static void main(String[] args) {
        // 包含重複元素的陣列
        int[] nums = { 1, 3, 6, 6, 6, 6, 6, 10, 12, 15 };
        System.out.println("\n陣列 nums = " + java.util.Arrays.toString(nums));

        // 二分搜尋左邊界和右邊界
        for (int target : new int[] { 6, 7 }) {
            int index = binarySearchLeftEdge(nums, target);
            System.out.println("最左一個元素 " + target + " 的索引為 " + index);
            index = binarySearchRightEdge(nums, target);
            System.out.println("最右一個元素 " + target + " 的索引為 " + index);
        }
    }
}

File: ja/codes/csharp/chapter_searching/binary_search_edge.cs
/**
* File: binary_search_edge.cs
* Created Time: 2023-08-06
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_searching;

public class binary_search_edge {
    /* 最も左の target を二分探索 */
    int BinarySearchLeftEdge(int[] nums, int target) {
        // target の挿入位置を探すのと等価
        int i = binary_search_insertion.BinarySearchInsertion(nums, target);
        // target が見つからなければ、-1 を返す
        if (i == nums.Length || nums[i] != target) {
            return -1;
        }
        // target が見つかったら、インデックス i を返す
        return i;
    }

    /* 最も右の target を二分探索 */
    int BinarySearchRightEdge(int[] nums, int target) {
        // 最左の target + 1 を探す問題に変換する
        int i = binary_search_insertion.BinarySearchInsertion(nums, target + 1);
        // j は最も右の target を指し、i は target より大きい最初の要素を指す
        int j = i - 1;
        // target が見つからなければ、-1 を返す
        if (j == -1 || nums[j] != target) {
            return -1;
        }
        // target が見つかったら、インデックス j を返す
        return j;
    }

    [Test]
    public void Test() {
        // 重複要素を含む配列
        int[] nums = [1, 3, 6, 6, 6, 6, 6, 10, 12, 15];
        Console.WriteLine("\n配列 nums = " + nums.PrintList());

        // 二分探索で左端と右端を探す
        foreach (int target in new int[] { 6, 7 }) {
            int index = BinarySearchLeftEdge(nums, target);
            Console.WriteLine("一番左の要素 " + target + " のインデックスは " + index);
            index = BinarySearchRightEdge(nums, target);
            Console.WriteLine("一番右の要素 " + target + " のインデックスは " + index);
        }
    }
}
